File: src/main/scala/lang/Ast.scala
package cql.lang

import io.circe.Encoder
import io.circe.syntax._
import io.circe.Json

trait Query

object QueryList {
  implicit val encoder: Encoder[QueryList] = Encoder.instance { list =>
    val arr = list.exprs.map {
      case q: CqlBinary         => q.asJson
      case q: QueryField          => q.asJson
      case q: QueryOutputModifier => q.asJson
    }
    Json.obj("type" -> "QueryList".asJson, "content" -> Json.arr(arr*))
  }
}
case class QueryList(
    exprs: List[CqlBinary | QueryField | QueryOutputModifier]
) extends Query

object CqlBinary {
  implicit val encoder: Encoder[CqlBinary] = Encoder.instance { queryBinary =>
    Json.obj(
      "type" -> "CqlBinary".asJson,
      ("left", queryBinary.left.asJson),
      ("right", queryBinary.right.asJson)
    )
  }
}
case class CqlBinary(
    left: QueryContent,
    right: Option[(Token, CqlBinary)] = None
)

object QueryContent {
  implicit val encoder: Encoder[QueryContent] = Encoder.instance {
    queryContent =>
      val content = queryContent.content match {
        case q: QueryStr    => q.asJson
        case q: CqlBinary => q.asJson
        case q: QueryGroup  => q.asJson
      }

      content.deepMerge(Json.obj("type" -> "QueryContent".asJson))
  }
}

case class QueryContent(content: QueryStr | CqlBinary | QueryGroup)

object QueryGroup {
  implicit val encoder: Encoder[QueryGroup] = Encoder.instance { group =>
    group.content.asJson.deepMerge(Json.obj("type" -> "QueryGroup".asJson))
  }
}

case class QueryGroup(content: CqlBinary)

object QueryStr {
  implicit val encoder: Encoder[QueryStr] = Encoder.instance { queryStr =>
    Json.obj(
      "type" -> "QueryStr".asJson,
      "searchExpr" -> queryStr.searchExpr.asJson
    )
  }
}

case class QueryStr(searchExpr: String) extends Query
object QueryField {
  implicit val encoder: Encoder[QueryField] = Encoder.instance { queryField =>
    Json.obj(
      "type" -> "QueryField".asJson,
      "key" -> queryField.key.asJson,
      "value" -> queryField.value.asJson
    )
  }
}

case class QueryField(key: Token, value: Option[Token]) extends Query

object QueryOutputModifier {
  implicit val encoder: Encoder[QueryOutputModifier] =
    Encoder.instance { queryField =>
      Json.obj(
        "type" -> "QueryOutputModifier".asJson,
        "key" -> queryField.key.asJson,
        "value" -> queryField.value.asJson
      )
    }
}

case class QueryOutputModifier(key: Token, value: Option[Token]) extends Query

File: prosemirror-client/src/lang/ast.ts
import { Token } from "./token";

export type QueryList = {
  type: "QueryList";
  content: QueryBinary[];
};

export const createQueryList = (content: QueryList["content"]): QueryList => ({
  type: "QueryList",
  content,
});

export type QueryBinary = {
  type: "QueryBinary";
  left: QueryContent;
  right?: [Token, QueryBinary];
};

export const createQueryBinary = (
  left: QueryBinary["left"],
  right?: QueryBinary["right"]
): QueryBinary => ({
  type: "QueryBinary",
  left,
  right,
});

export type QueryContent = {
  type: "QueryContent";
  content: QueryStr | QueryBinary | QueryGroup | QueryField;
};

export const createQueryContent = (
  content: QueryContent["content"]
): QueryContent => ({
  type: "QueryContent",
  content,
});

export type QueryGroup = { type: "QueryGroup"; content: QueryBinary };

export const createQueryGroup = (
  content: QueryGroup["content"]
): QueryGroup => ({
  type: "QueryGroup",
  content,
});

export type QueryStr = { type: "QueryStr"; searchExpr: string; token: Token };

export const createQueryStr = (token: Token): QueryStr => ({
  type: "QueryStr",
  searchExpr: token.literal ?? "",
  token,
});

export type QueryField = { type: "QueryField"; key: Token; value?: Token };

export const createQueryField = (
  key: QueryField["key"],
  value: QueryField["value"]
): QueryField => ({
  type: "QueryField",
  key,
  value,
});
